### DISCUSSION TITLE
`await` call in `run_in_threadpool` taking 50% of total request time.

### DISCUSSION THREAD

**Discussion by ayomide865** [fastapi/fastapi#13065] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# The application is a standard FastAPI application, taken from the docs.
# No special features are being used. 
# Endpoints are defined with `async` but the code is actually sync
# Running with uvicorn workers: `poetry run uvicorn server.main:app --workers 8`
# Again, we don't have a problem, but just want clarification on the long `await` time

from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def api():
    return {"hello": "world"}
```

### Description

Hello!

We don't really have a problem; we just want some clarification on something we observed on one of our endpoints.

We used `pyinstrument` to visualize the span of one of the requests to the application, and we noticed that the total request time was ~600ms, with the `run_in_threadpool` call -- specifically the `await` -- taking ~350ms (>50% of the total request time). So, we're curious how we could reduce that time. 

All of our endpoints are sync, so we expect some blocking, hence the running in threadpool. However, the actual endpoint function takes < 30ms to complete, and pydantic response serialization contributes to the remaining ~300ms time (both of which are fine). Our main concern is the very long `await` time at the beginning of the request.

So, what really is the `await` call, and why is it taking so long? Is this the time it takes for one of the `uvicorn`/`gunicorn` workers to pick up the task or the time it takes for `anyio` to spin up the worker thread? These are our guesses, we're probably wrong! We tested this on a local machine with reasonable specs so we don't expect the CPU/Hardware to be the bottleneck.

The timing of the `await` call sometimes varies but it always, usually takes a significant chunk of the total request time and always at the beginning of the request (according to `pyinstrument`). We were hoping to get some insights on this, especially as it's a call that has nothing to do with the endpoint function itself. 

Thank you.

![Screenshot from 2024-12-12 14-43-54](https://github.com/user-attachments/assets/53c9dd33-123c-45d1-8f5b-a6e87ef8c663)

Same thing here:
![Screenshot from 2024-12-12 15-56-07](https://github.com/user-attachments/assets/bc4f672c-a5c0-4864-bccf-3682f33173b4)

### Operating System

Linux

### Operating System Details

Distro: Pop!_OS 22.04 LTS x86_64 
CPU: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics (16) @ 3.300GHz 
Memory: 32GB

### FastAPI Version

0.113.0

### Pydantic Version

2.9.0

### Python Version

Python 3.10.12

**Comment by YuriiMotov** [comment]
Your question looks quite confusing.. You shared a code example, but it looks like you are using another code to run with pyinstruments.

I don't see anything suspicious on the screenshots.
You app is async and to run sync endpoints it uses threadpool. It creates an async task what waits until function you run in the threadpool has been done.

I think we could give you more details and probably recommendations if you created a simple MRE.

**Reply by ayomide865** [reply]
Hey, thanks for responding. 

I already mentioned that the code has nothing to do with the question I'm asking, hence my sharing a boilerplate code sample.

My question really is why the `await` time **at the beginning of the request** takes so long. You can see in the screenshot that the actual request handling, which is run in the `run_endpoint_function` is called **after** the long `await` call. My question, again, is why does it take so long for that `await` to finish if FastAPI hasn't even started handling the request.

You'll also see that the requests themselves take very little time, e.g. in the second screenshot. If, indeed, a thread was spawned at the beginning to handle the request, shouldn't it finish around the same time the request is finishing? In the same screenshot, the entire `run_endpoint_function` takes <20ms, but the `await` takes ~400ms. Wouldn't the thread finish its `await`ing after a little over 20ms? 

This discrepancy is what I don't understand, and is what I'm trying to clarify. 400ms is a lot of idle latency and I'm just trying to understand where it is coming from. All the latency in the application is coming from these randomly long `await`s.

**Comment by yakubka** [comment]
The overhead you are measuring is mostly thread pool scheduling, not the `await` itself. `run_in_threadpool` submits a callable to `anyio`s default thread executor and then suspends the calling coroutine until the thread finishes. Each submission involves acquiring the thread pool lock, posting to the work queue, and context-switching back to the event loop after the thread completes. On a busy event loop that context switch alone can account for several hundred microseconds.

If the function being offloaded is very fast (sub-millisecond), the scheduling overhead will dominate and you will see the pattern you described.

The fix is to only use `run_in_threadpool` for genuinely blocking I/O or CPU-bound work that takes long enough to justify the overhead. For fast synchronous utilities, calling them directly from an `async def` handler is fine as long as they do not block.

```python
# no benefit if compute_thing() takes < 1ms
result = await run_in_threadpool(compute_thing, arg)

# just call it directly instead
result = compute_thing(arg)
```

If the function does block and you cannot make it async, the overhead is unavoidable, but you can reduce it by batching work into fewer threadpool calls rather than calling `run_in_threadpool` many times in a loop.
